### DISCUSSION TITLE
Forked repo's commits are ahead of the base repo.

### DISCUSSION THREAD

**Discussion by Kalaiselvi84** [community/community#58914] [Repositories] [answered]
### Select Topic Area

Question

### Body

Hi Team, 

I have a [forked repo](https://github.com/Kalaiselvi84/agones) and [this](https://github.com/googleforgames/agones) is the base repo. 

I am trying to sync my fork with base repo through GUI/CLI. Both logs are suggesting to use the pull request but there are 0 files changed. When I click on `Update Branch` in Github, the new commits are getting added to the ahead of commits list and if I click on `Discard x commits` it says `protected branch 'main' check failed: Changes must be made through a pull request. Cannot force-push to this branch`. 

Please let me know how to resolve this issue.

Thanks!

**Comment by airtower-luna** [comment]
The error message sounds like you have a branch protection rule for the `main` branch in your fork. You'll need to disable that if you want to overwrite it with the upstream version.

**Comment by Kalaiselvi84** [comment]
I have disabled the branch protection for the `main` branch and ran following commands,

$ git fetch upstream
$ git checkout main
$ git merge upstream/main

$ git push origin main
Enumerating objects: 1, done.
Counting objects: 100% (1/1), done.
Writing objects: 100% (1/1), 244 bytes | 244.00 KiB/s, done.
Total 1 (delta 0), reused 0 (delta 0), pack-reused 0
To https://github.com/Kalaiselvi84/agones.git
   065359748..837275954  main -> main

If you see in [my fork](https://github.com/Kalaiselvi84/agones) one more commit got added into the ahead of commits list. Please share me the correct commands to fix this issue.

**Comment by airtower-luna** [comment]
It seems like you're merging with merge commits. If you want the `main` branch in your fork to be the same as upstream you'll have to [*reset*](https://git-scm.com/docs/git-reset) it.

**Warning:** That reset will *permanently discard* any commits you have added to your `main` branch both locally and in your fork on GitHub (unless you also have them on other branches). Don't do it if that's not what you want, and I highly recommend to make sure your working tree is clean first to avoid any confusion. :warning: 

```
git checkout main
git reset --soft upstream/main
git push --force origin
```

Two more notes:

1. [`git reset --soft`](https://git-scm.com/docs/git-reset#Documentation/git-reset.txt---soft) resets only the current branch, not your index or working tree. You might see *a lot* of changed files, or not, depending on how different the branch is after. If you want to overwrite your working tree and index too, use `--hard` (do that only if you're sure – there's no way to recover uncommitted changes). :warning: 
2. When you want to update your `main` branch in the future, make sure to do fast-forward merges *only*, otherwise you'll end up with merge commits again.

**Reply by Kalaiselvi84** [reply]
```
$ git checkout main
Already on 'main'
Your branch is up to date with 'origin/main'.
$ git reset --soft upstream/main
$ git push --force origin
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
remote: error: GH006: Protected branch update failed for refs/heads/main.
remote: error: Cannot force-push to this branch
To https://github.com/Kalaiselvi84/agones.git
 ! [remote rejected]     main -> main (protected branch hook declined)
error: failed to push some refs to 'https://github.com/Kalaiselvi84/agones.git'
```

but I have removed all the branch protection in main branch in my forked repo - https://screenshot.googleplex.com/8ZeRPvvyqyudaLf

**Reply by airtower-luna** [reply]
GitHub seems to think otherwise. Mind that there currently are *branch protection rules* and (separately!) *repository rules* in beta. :sweat_smile: 

Also I can's see that screenshot it leads to something that might be a Google login thing, but looks a bit suspicious.

**Reply by Kalaiselvi84** [reply]
<img width="778" alt="Screenshot 2023-06-23 at 1 57 54 PM" src="https://github.com/community/community/assets/117940852/858dce2d-6295-47bd-85b2-967e3e1546cb">

I tried with --hard as well

```
$ git reset --hard upstream/main
HEAD is now at ee0062012 Potential fix for TestAllocatorAllocate* flakiness (#3232)
$ git push --force origin
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
remote: error: GH006: Protected branch update failed for refs/heads/main.
remote: error: Cannot force-push to this branch
To https://github.com/Kalaiselvi84/agones.git
 ! [remote rejected]     main -> main (protected branch hook declined)
error: failed to push some refs to 'https://github.com/Kalaiselvi84/agones.git'
```

**Reply by airtower-luna** [reply, marked as answer]
That still *is* a branch protection rule, you'd need to use the "delete" button on this screen:

![grafik](https://github.com/community/community/assets/5814919/7859c48b-3f2e-4ba6-b072-c6f0ec9c3027)

Whether you do a soft or hard reset changes only how your *local* data is treated, it doesn't matter for pushing.

**Reply by Kalaiselvi84** [reply]
It worked, thanks a ton!

**Comment by nuri-yoo** [comment]
To quickly check whether a forked repo is ahead or behind the base, I built [branch-watch](https://github.com/nuri-yoo/branch-watch). `bw forks` shows ahead/behind commit counts for all your forks vs. upstream at a glance from the terminal.
